proposals/nnnn-span-access-shared-contiguous-storage.md

```diff
@@ -0,0 +1,821 @@
+
+##### Use a non-escapable index type
+A non-escapable index type implies that any indexing operation would borrow its `Span`. This would prevent using such an index for a mutation, since a mutation requires an _exclusive_ borrow. Noting that the usage pattern we desire for `Span` must also apply to `MutableSpan`(described [below](#MutableSpan),) a non-escapable index would make it impossible to also implement a mutating subscript, unless any mutating operation consumes the index. This seems untenable.
+
+##### Naming
+
+The ideas in this proposal previously used the name `BufferView`. While the use of the word "buffer" would be consistent with the `UnsafeBufferPointer` type, it is nevertheless not a great name, since "buffer" is commonly used in reference to transient storage. Another previous pitch used the term `StorageView` in reference to the `withContiguousStorageIfAvailable()` standard library function. We also considered the name `StorageSpan`, but that did not add much beyond the shorter name `Span`. `Span` clearly identifies itself as a relative of C++'s `std::span`.
+
+##### A more sophisticated approach to indexing
+
+This is discussed more fully in the [indexing appendix](#Indexing) below.
+
+## <a name="Directions"></a>Future directions
+
+#### <a name="Initializers"></a>Initializing and returning `Span` instances
+
+A `Span` represents a region of memory and, as such, must be initialized using an unsafe pointer. This is an unsafe operation which will typically be performed internally to a container's implementation. In order to bridge to safe code, these initializers require new annotations that indicate to the compiler how the newly-created `Span` can be used safely.
+
+These annotations have been [pitched][PR-2305-pitch] and are expected to be formally [proposed][PR-2305] soon. `Span` initializers using lifetime annotations will be proposed alongside the annotations themselves.
+
+#### Obtaining variant `Span`s and `RawSpan`s from `Span` and `RawSpan`
+
+`Span`s representing subsets of consecutive elements could be extracted out of a larger `Span` with an API similar to the `extracting()` functions recently added to `UnsafeBufferPointer` in support of non-copyable elements:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public func extracting(_ bounds: Range<Int>) -> Self
+}
+```
+
+Each variant of such a function needs to return a `Span<Element>`, which requires a lifetime dependency. 
+
+Similarly, a `RawSpan` should be initializable from a `Span<T: BitwiseCopyable>`, and `RawSpan` should provide a function to unsafely view its content as a typed `Span`:
+
+```swift
+extension RawSpan {
+  public init<T: BitwiseCopyable>(_ span: Span<T>)
+
+  public func unsafeView<T: BitwiseCopyable>(as type: T.Type) -> Span<T>
```

Comment:
Since you're mentioning `RawSpan.unsafeView<T: BitwiseCopyable>(as type: T.Type) -> Span<T>` here, we may also want to explain that this is a safe API as long as the memory contains a value bit pattern for `T` at each position. `Span<T>` does not impose any strict aliasing demands on the programmer; they can use multiple read-only spans to safely view the same underlying memory as different types.